pkg/engine/lifecycletest/test_plan.go

```diff
@@ -431,3 +433,53 @@ func MakeBasicLifecycleSteps(t *testing.T, resCount int) []TestStep {
 		},
 	}
 }
+
+type TestBuilder struct {
+	t       *testing.T
+	loaders []*deploytest.ProviderLoader
+	snap    *deploy.Snapshot
+}
+
+func NewTestBuilder(t *testing.T, snap *deploy.Snapshot) *TestBuilder {
+	return &TestBuilder{
+		t:       t,
+		snap:    snap,
+		loaders: make([]*deploytest.ProviderLoader, 0, 1),
```

Comment:
Nit: Should we use our [`slice.Prealloc`](https://github.com/pulumi/pulumi/blob/master/sdk/go/common/slice/slice.go) helper?

```suggestion
		loaders: slice.Prealloc[*deploytest.ProviderLoader](1)
```